### PR TITLE
adjust Scheduler metrics

### PR DESCRIPTION
**Pull request by UncP**
`KV_COMMAND_GC_SKIPPED_COUNTER`
`KV_COMMAND_GC_EMPTY_RANGE_COUNTER`
`KV_COMMAND_KEYREAD_HISTOGRAM_VEC`

Above are the metrics that changed in thread pool, so we create a local copy for each of them.

### CODE REVIEW

File: src/storage/metrics.rs [hunk 1]
```diff
@@ -137,3 +137,29 @@ lazy_static! {
             exponential_buckets(1.0, 2.0, 21).unwrap()
         ).unwrap();
 }
+
+pub struct LocalCounter {
```
Comment (siddontang) [inline src/storage/metrics.rs]: I suggest moving this to the util so other mod can also use this.

File: src/storage/metrics.rs [hunk 2]
```diff
@@ -137,3 +137,29 @@ lazy_static! {
...
+    /// `inc` increments the local counter by 1.
+    #[inline]
+    pub fn inc(&mut self) {
+        self.val += 1.0;
+    }
+
+    /// `flush` the local counter value to the counter
+    #[inline]
+    pub fn flush(&self) -> Result<()> {
+        self.counter.inc_by(self.val)
```
Comment (siddontang) [inline src/storage/metrics.rs]: After flush, we should clean up the value. 
No need to return Result, using unwrap here is fine.

Comment (siddontang) [reply src/storage/metrics.rs]: if val is 0, no need to flush to global counter.

File: src/storage/metrics.rs (hunk 1 above)
Comment (siddontang) [inline src/storage/metrics.rs]: I suggest moving this to the util so other mod can also use this.

File: src/storage/txn/scheduler.rs [hunk 3]
```diff
@@ -1125,7 +1144,7 @@ impl Scheduler {
                     .with_label_values(&[tag])
                     .start_coarse_timer();
 
-                let s = process_read(cid, cmd, ch, snapshot);
+                let s = process_read(cid, cmd, ch, snapshot, ctx);
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: Mostly, ctx should be the first argument.

File: src/storage/txn/scheduler.rs [hunk 4]
```diff
@@ -416,6 +416,7 @@ fn process_read(
     mut cmd: Command,
     ch: SyncSendCh<Msg>,
     snapshot: Box<Snapshot>,
+    sched_ctx: &mut SchedContext,
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: why not handle process_write at same time?

**Comment by siddontang** [comment]
any benchmark?

File: src/storage/txn/scheduler.rs (hunk 4 above)
Comment (UncP) [reply src/storage/txn/scheduler.rs]: Because there is nothing need to be local in process_write, only these 3 metrics need to be local.

**Comment by huachaohuang** [comment]
@siddontang @overvenus 
Prometheus has a `LocalHistogram`, should we move the `LocalCounter` to there too?

**Comment by UncP** [comment]
I think it's much better.

**Comment by siddontang** [comment]
Good idea, @overvenus

**Comment by UncP** [comment]
/| master | branch
--|-----|--
99% get avg duration(ms) | 22.3 | 22.4|
95% get avg duration(ms) | 14.0| 13.4|
get ops | 81.1k | 81.6k |

I am gonna create a pr in `Prometheus`.

**Comment by UncP** [comment]
@zhangjinpeng1987 @huachaohuang PTAL

**Comment by huachaohuang** [comment]
LGTM

**Pull request merged**